### PR TITLE
flux-jobs: add --from-stdin option and other small fixes

### PR DESCRIPTION
**Pull request by grondo**
This PR started as an implementation of #2638 in order to add a sane test for #2634 which required a fix for #2281 

Some other small fixes came along way:
 - Moved the `CleanFormatter` argparse `HelpFormatter` class to `flux.util.help_formatter` and added it to existing python flux commands for consistent `--help` output.
 - Added a rewrite of the generic  error thrown in `flux.Flux()` to include `Unable to connect to Flux`, improving the default error message from flux python commands when they can't connect to flux (from `Errno 2: ENOENT: No such file or directory` to `Unable to connect to Flux: ENOENT: No such file or directory` -- a slight improvement)

Additionally a framework for adding tests was added to `t2800-jobs-cmd.t`. Just create a directory under `t/flux-jobs/tests/` and throw in some input, expected output, an optional description and format. Hopefully it comes in useful (but maybe there will only ever be one test under there, oh well)

### CODE REVIEW

**Comment by grondo** [comment]
LGTM found a couple good python boo boos, so I just force-pushed fixes for those.

**Comment by chu11** [comment]
Overall LGTM, nothing stood out except no test for #2644 using the new `--form-stdin`?

**Comment by grondo** [comment]
> Overall LGTM, nothing stood out except no test for #2644 using the new --form-stdin?

Hm, I hadn't even thought of that. The new test for #2634 uses that option, but I wonder what a standalone test for that feature would look like.

I'm also missing a test for the fallback to userid when `getpwuid` fails. However, finding a userid that didn't exist on a system seemed finicky enough that I bailed on that one too. However, it wouldn't be impossible to make a reliable test (iterate userids maybe until one fails, then use that userid for the input for the test?)

**Comment by chu11** [comment]
> Hm, I hadn't even thought of that. The new test for #2634 uses that option, but I wonder what a standalone test for that feature would look like.

I thought testing via `--from-stdin` made sense to "fake" a bad ID b/c I don't think we can submit a job using a fake job id?

Hadn't thought about how to ensure a bad ID is truly bad.  I had just assumed lets pick something like 65536.

**Comment by grondo** [comment]
Restarted the ASan builder (#2651)

**Comment by grondo** [comment]
> Hadn't thought about how to ensure a bad ID is truly bad. I had just assumed lets pick something like 65536.

Yeah, but that userid could exist on some random system so `make check` would fail there.
(Apologies, I misread your comment at first and thought you were suggesting there was no test for `--from-stdin`)

I'll attempt to make a reliable test by searching for a bad userid by running `getpwuid()` until it fails.

File: src/bindings/python/flux/core/handle.py [hunk 1]
```diff
@@ -38,7 +38,12 @@ def __init__(self, url=ffi.NULL, flags=0, handle=None):
         )
 
         if handle is None:
-            self.handle = raw.flux_open(url, flags)
+            try:
+                self.handle = raw.flux_open(url, flags)
+            except Exception as err:
+                raise EnvironmentError(
+                    err.errno, "Unable to connect to Flux: {}".format(err.strerror)
+                )
```
Comment (SteVwonder) [inline src/bindings/python/flux/core/handle.py]: Generic `Exception`s aren't guaranteed to have an `errno` field.  Only `EnvironmentError`s (and by proxy `IOError`s and `OSError`s) are documented to have one [(docs link)](https://docs.python.org/2/library/exceptions.html#exceptions.EnvironmentError).  So this exception handle can potentially cause it's own exception on the `err.errno` access.  I recommend limiting the scope of the exceptions caught from `Exception` to just `EnvironmentError`.

File: src/cmd/flux-jobs.py [hunk 2]
```diff
@@ -89,54 +96,28 @@ def output_format(fmt, jobs):
...
+    ignored.
+    """
+    import fileinput
+    import json
+
+    jobs = []
+    for line in fileinput.input("-"):
+        try:
+            job = json.loads(line)
+        except Exception as e:
```
Comment (SteVwonder) [inline src/cmd/flux-jobs.py]: I'm surprised pylint didn't complain about this broad exception block.  Since the next action is to log an error and exit, it is actually probably fine to leave this the way it is.   FWIW (since I already looked it up), the py3 docs say that this method will throw `JSONDecodeError` [(source)](https://docs.python.org/3/library/json.html#json.loads).  Feel free to leave this the way it is though.

File: src/bindings/python/flux/core/handle.py (hunk 1 above)
Comment (grondo) [reply src/bindings/python/flux/core/handle.py]: Ok great! I really needed help with this code. For the life of me I couldn't find documentation on what functions throw which exceptions. Thanks!

File: src/cmd/flux-jobs.py (hunk 2 above)
Comment (grondo) [reply src/cmd/flux-jobs.py]: I don't think we're running pylint on commands :rofl: - so I should probably fix this!

File: src/bindings/python/flux/core/handle.py (hunk 1 above)
Comment (grondo) [reply src/bindings/python/flux/core/handle.py]: @SteVwonder is it ok to rethrow `EnvironmentError` in this case? Can you think of a better way to generally improve the error message from `flux.Flux()`?

Comment (SteVwonder) [reply src/bindings/python/flux/core/handle.py]: > @SteVwonder is it ok to rethrow EnvironmentError in this case? 

Yeah, I think that is totally reasonable.

> For the life of me I couldn't find documentation on what functions throw which exceptions. Thanks!

Yeah....that's definitely a big problem (in Python and C++).  Exception are great and yet still terrible.  Unfortunately, I don't have a better answer other than looking at the Python Docs (which can be a crap shot), and learning the hard way by shooting yourself in the foot :gun:.

File: src/cmd/flux-jobs.py (hunk 2 above)
Comment (grondo) [reply src/cmd/flux-jobs.py]: Would it be acceptable to catch `ValueError` instead?

Comment (chu11) [reply src/cmd/flux-jobs.py]: as an aside since you're here, I think `lgtm` will flag the single character variable `e`.

**Comment by grondo** [comment]
Ok, I've pushed some fixes to address comments (still need to squash them if they look ok)

 - added corner-case tests for `flux jobs --from-stdin` (no input, invalid input)
 - added a test for invalid userid by searching for first invalid userid on the current system
 - Handle `EnvironmentError` exception explicitly in `flux.Flux()` instead of generic exception

I also threw on a couple of other very minor tweaks for flux-jobs output (I'm fine removing them if they are too much)

 - Format durations < 10s > 1.e-3 with "%.3fs" instead of "%.2g" (improves consistency of output for small runtimes)
 - Rename "TIME" header to "RUNTIME", since that is what it is, and we may have other *TIME output in the future.

File: src/cmd/flux-jobs.py (hunk 2 above)
Comment (grondo) [reply src/cmd/flux-jobs.py]: pylint, yeah. Though its annoying I'll fix that too.

File: src/cmd/flux-jobs.py [hunk 3]
```diff
@@ -42,7 +42,9 @@ def runtime(job, roundup):
 
 def runtime_fsd(job, hyphenifzero):
     t = runtime(job, False)
-    if t < 60.0:
+    if t >= 1e-3 and t < 10.0:
```
Comment (chu11) [inline src/cmd/flux-jobs.py]: I think there's a logic hole for `< 1e-3`.  Do we just want to round down and call it `0s`?

Comment (chu11) [reply src/cmd/flux-jobs.py]: oh wait, nevermind, it'll fall through to `else:` and `0s`.

Comment (chu11) [reply src/cmd/flux-jobs.py]: no, i'm getting my functions mixed up, the fallthrough would be the `days` else case.

**Comment by grondo** [comment]
Force pushed and squashed up all the fixups. 

@chu11, I think I addressed your comments above, but better review again because I worked a bit quickly this AM before leaving for a school function.

**Comment by chu11** [comment]
LGTM

**Comment by grondo** [comment]
I had to force-push again to make the python formatter happy.

**Comment by grondo** [comment]
Just curious: @SteVwonder @trws, @chu11 any idea why lgtm marks [this code](https://lgtm.com/projects/g/flux-framework/flux-core/snapshot/9ebbd74f28f6f8f667539a964452ac047fa28a29/files/src/bindings/python/flux/util.py?sort=name&dir=ASC&mode=heatmap#xc0a0dc532970d9e8:1) unreachable?

**Comment by trws** [comment]
That's some surprisingly good analysis... Since the try block only contains a return of an unevaluated lambda, the only way that could fail would be perhaps an allocation error.  The FluxHelpFormatter is never actually constructed there, it just captures the upvals and returns the new callable, so there's nowhere for a TypeError to come from.

**Comment by grondo** [comment]
Thanks, maybe I should fix that here, though this PR only moves the code (the incorrect code hails from StackOverflow, and is meant to fall back to a meaningful error if the internals of argparse ever change. It would probably be sufficient to just let any exception take the program down)

**Comment by trws** [comment]
That would be ok, or maybe you want that try/except inside the lambda (which would require making it a local function) which would then do something sensible if the instantiation fails?

**Comment by grondo** [comment]
> That would be ok, or maybe you want that try/except inside the lambda (which would require making it a local function) which would then do something sensible if the instantiation fails?

Ehh, I played around with this and any exception due to internal changes in `argparse` incompatible with our HelpFormatter aren't thrown until after parse_args() is called. So, any try/catch in `flux.util.help_formatter()` is too early.

@trws, what do you think of just manually checking the `argparse.HelpFormatter` class for the members our subclass requires, and falling back with a warning if any aren't found, e.g.:

```diff
diff --git a/src/bindings/python/flux/util.py b/src/bindings/python/flux/util.py
index d09aa89c6..bf2291340 100644
--- a/src/bindings/python/flux/util.py
+++ b/src/bindings/python/flux/util.py
@@ -80,6 +80,14 @@ def help_formatter(argwidth=40):
     Return our 'clean' HelpFormatter, if possible, with a wider default
      for the max width allowed for options.
     """
+    import inspect
+    import warnings
+
+    required = ("_format_action_invocation", "_metavar_formatter", "_format_args")
+    available = dir(argparse.HelpFormatter)
+    if not all(name in available for name in required):
+        warnings.warn("argparse methods have changed, falling back.")
+        return lambda prog: argparse.HelpFormatter(prog, max_help_position=argwidth)
 
     class FluxHelpFormatter(argparse.HelpFormatter):
         def _format_action_invocation(self, action):
@@ -107,16 +115,7 @@ def help_formatter(argwidth=40):
             args_string = self._format_args(action, default)
             return optstring + "=" + args_string
 
-    try:
-        # https://stackoverflow.com/a/5464440
-        # beware: "Only the name of this class is considered a public API."
-        # https://stackoverflow.com/questions/44333577
-        return lambda prog: FluxHelpFormatter(prog, max_help_position=argwidth)
-    except TypeError:
-        import warnings
-
-        warnings.warn("argparse help formatter failed, falling back.")
-        return argparse.HelpFormatter
+    return lambda prog: FluxHelpFormatter(prog, max_help_position=argwidth)
 
 
 class CLIMain(object):

```

**Comment by SteVwonder** [comment]
> Ehh, I played around with this and any exception due to internal changes in argparse incompatible with our HelpFormatter aren't thrown until after parse_args() is called. So, any try/catch in flux.util.help_formatter() is too early.

Ouch!  That's unfortunate.

> what do you think of just manually checking the argparse.HelpFormatter class for the members our subclass requires, and falling back with a warning if any aren't found

Seems fine to me.  My only comment would be that `hasattr` might be a better way to check for the member existance rather than `dir` [(relevant StackOverflow post)](https://stackoverflow.com/questions/17723569/whats-the-difference-between-hasattr-and-attribute-in-dir).

**Comment by grondo** [comment]
Thanks! I somehow didn't think getattr would work on a class, but it sure does!

**Comment by grondo** [comment]
New code:
```python
    import warnings

    required = ("_format_action_invocation", "_metavar_formatter", "_format_args")
    if not all(hasattr(argparse.HelpFormatter, name) for name in required):
        warnings.warn("argparse methods have changed, falling back.")
        return lambda prog: argparse.HelpFormatter(prog, max_help_position=argwidth)
```

Sanity check (modified one of the required methods to something not in the class)
Before:
```console
$ FLUX_PYCLI_LOGLEVEL=10 src/cmd/flux jobs -h 
usage: flux-jobs [-h] [-a] [-A] [-c N] [-s STATES] [--suppress-header]
                 [-u [USERNAME|UID]] [-o FORMAT]

optional arguments:
  -h, --help                 show this help message and exit
  -a                         List all jobs for current user
  -A                         List all jobs for all users
  -c, --count=N              Limit output to N jobs(default 1000)
  -s, --states=STATES        List jobs in specific
                             states(pending,running,inactive)
      --suppress-header      Suppress printing of header line
  -u, --user=[USERNAME|UID]  Limit output to specific username or userid
                             (Specify "all" for all users)
  -o, --format=FORMAT        Specify output format using Python's string
                             format syntax
```
```console 
$ src/cmd/flux jobs -h 
/home/grondo/git/f.git/src/bindings/python/flux/util.py:88: UserWarning: argparse methods have changed, falling back.
  warnings.warn("argparse methods have changed, falling back.")
usage: flux-jobs [-h] [-a] [-A] [-c N] [-s STATES] [--suppress-header]
                 [-u [USERNAME|UID]] [-o FORMAT]

optional arguments:
  -h, --help                            show this help message and exit
  -a                                    List all jobs for current user
  -A                                    List all jobs for all users
  -c N, --count N                       Limit output to N jobs(default 1000)
  -s STATES, --states STATES            List jobs in specific
                                        states(pending,running,inactive)
  --suppress-header                     Suppress printing of header line
  -u [USERNAME|UID], --user [USERNAME|UID]
                                        Limit output to specific username or
                                        userid (Specify "all" for all users)
  -o FORMAT, --format FORMAT            Specify output format using Python's
                                        string format syntax
```

**Comment by grondo** [comment]
BTW, is `warnings.warn` the right thing to use here? It has that weird double print thing going on?

**Comment by SteVwonder** [comment]
> BTW, is warnings.warn the right thing to use here? It has that weird double print thing going on?

The other option (that I'm aware of) would be to use `logging.warn`.  I wasn't actually sure of the differences, but [this SO post was pretty helpful](https://stackoverflow.com/a/14762106).  I don't have a strong opinion one way or the other.  So I guess try out `logging.warn` and if you like the output better, then let's go with that.

**Comment by SteVwonder** [comment]
>  1 for Unused named argument in formatting call

Wow! This LGTM bot is doing a lot behind the scene.  It noticed that our format string (which impressively is not in the same function) doesn't contain certain keywords, and so passing those keywords to the format method is not useful.  Unfortunately, it didn't catch that there is a whole other code path involved user-provided input.  Filed a bug upstream for it: https://github.com/Semmle/ql/issues/2650

If you want to silence [the error](https://lgtm.com/projects/g/flux-framework/flux-core/snapshot/ee30ed85d95632cea4f1bb1a1a4deb1046fd95b5/files/src/cmd/flux-jobs.py#xc23648d36039a520:1), you can stick `lgtm [py/str-format/surplus-named-argument]` at the end of the `fmt.format` line (81).

**Comment by grondo** [comment]
Oh yeah, I prefer this:
```python
        logging.getLogger(__name__).warning(
            "required argparse methods missing, falling back to HelpFormatter."
        )
```

Which prints:
```
flux.util: WARNING: required argparse methods missing, falling back to HelpFormatter.
```

**Comment by grondo** [comment]
Just pushed a test and a fix for #2658.

Since the number of fixes collecting in this PR is getting kind of large, I'd be fine splitting up into separate features/fixes PRs if requested.

**Comment by chu11** [comment]
I think this is ready to go, you think its ready @grondo?

**Comment by grondo** [comment]
Sure, thanks

**Comment by grondo** [comment]
Thanks!
